listener.go

```diff
@@ -74,6 +74,20 @@ func (l *listener) Accept() (tpt.CapableConn, error) {
 			sess.CloseWithError(errorCodeConnectionGating, "connection gated")
 			continue
 		}
+
+		// return through active hole punching if any
+		hpkey := sess.RemoteAddr().String()
+		l.transport.holePunchingMx.Lock()
+		holePunch, ok := l.transport.holePunching[hpkey]
+		l.transport.holePunchingMx.Unlock()
+		if ok {
+			select {
```

Comment:
There's a race condition here: If we receive 2 connections before the other go routine has deleted this entry from the `holePunching` map, we'll block here until the context is done.

Reply:
1. It's fixed now, right?
2. Why not delete from the hole-punching map when we first access it? That should fix all the problems here, right?

Reply:
Actually, if we one-buffer the channel and atomically delete from the map, we _should_ be able to do everything under the same lock:

```go
		l.transport.holePunchingMx.Lock()
		holePunch, ok := l.transport.holePunching[key]
        if ok {
            holePunch.connCh <- conn
            delete(l.transport.holePunching, key)
        }
		l.transport.holePunchingMx.Unlock()
```

Probably not necessary, but may make things easier to reason about.